perlxs_derive/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+        syn::Body::Struct(ref vdata) => {
+            match vdata {
+                &VariantData::Struct(ref fields) => {
+                    internals::ast::fields_from_ast(&errors, fields)
+                },
+                &VariantData::Tuple(_) | &VariantData::Unit => {
+                    panic!("You can only derive this for normal structs!");
+                },
+            }
+        },
+        syn::Body::Enum(_) => panic!("You can only derive this on structs!"),
+    };
+
+    errors.check().unwrap();
+
+    let mut letvars = Vec::new();
+    let mut matchparts = Vec::new();
+    let mut paramtests = Vec::new();
+    let mut paramvars  = Vec::new();
+
+    for field in fields.iter(){
+        let ident    = &field.ident;
+        let ty       = &field.ty;
+        let var      = Ident::new(format!("value_{}", ident));
+        let ty_lit   = Lit::Str(quote!{#ty}.to_string(),StrStyle::Cooked);
+
+        #[allow(unused_variables)]
+        let keys_lit : Vec<_> = field.keys.iter().map(|k| Lit::Str(k.to_string(),StrStyle::Cooked) ).collect();
+
+        letvars.push(
+            quote! {
+                let mut #var : Option<#ty> = None
+            }
+        );
+
+        for key in field.keys.iter(){
+            let key_lit   = Lit::Str(key.to_string(),StrStyle::Cooked);
+
+            matchparts.push(quote!{
+                #key_lit => {
+                    match ctx.st_try_fetch::<#ty>(i+1) {
+                        Some(Ok(v))  => {
+                            #var = Some( v );
+                        },
+                        Some(Err(e)) => {
+                            errors.push(_perlxs::error::ToStructErrPart::ValueParseFail{key: #key_lit, ty: #ty_lit, error: e.to_string(), offset: i+1});
+                        },
+                        None         => {
+                            errors.push(_perlxs::error::ToStructErrPart::OmittedValue(#key_lit));
+                        },
+                    }
+                }
+            });
+        }
+
+        if field.optional {
+            paramvars.push(
+                quote! {
+                    #ident: #var
+                }
+            );
+        }else{
+            paramtests.push(
+                quote!{
+                    if #var.is_none() {
+                        errors.push(_perlxs::error::ToStructErrPart::OmittedKey(&[#(#keys_lit),*]));
+                    };
+                }
+            );
+
+            paramvars.push(
+                quote! {
+                    #ident: #var.unwrap()
+                }
+            );
+        }
+
+    }
+
+    let from_kv_stack = quote!{
+
+        let mut i = offset;
+        while let Some(sv_res) = ctx.st_try_fetch::<String>(i) {
+            match sv_res {
+                Ok(key) => {
+                    match &*key {
+                        #(#matchparts,)*
+                        &_ => {
+                            // Unknown key. Should we warn?
```

Comment:
A warning here would make sense I think, but perl-xs lacks bindings for that yet. I am okay with leaving it as is for now and revisiting later, when `warn` is properly implemented.